torch/_dynamo/variables/builtin.py

```diff
@@ -1209,13 +1210,56 @@ def call_setattr(
             tx.output.side_effects.is_attribute_mutation(obj)
             and name_var.is_python_constant()
         ):
-            name = name_var.as_python_constant()
-            if name == "requires_grad" and isinstance(obj, variables.TensorVariable):
-                unimplemented(
-                    "mutating requires_grad can introduce a new leaf from non-leaf or vice versa in "
-                    "the middle of the graph, which aot_autograd does not currently know how to handle. "
-                )
-            tx.output.side_effects.store_attr(obj, name, val)
+            if isinstance(obj, variables.TensorVariable):
+                from .builder import wrap_fx_proxy
+
+                name = name_var.as_python_constant()
+                if name == "requires_grad":
+                    # TODO(voz): Make it work properly
+                    unimplemented(
+                        "mutating requires_grad can introduce a new leaf from non-leaf or vice versa in "
+                        "the middle of the graph, which aot_autograd does not currently know how to handle. "
+                    )
+                if name == "data":
+                    # Remove the old reference in tracked fakes - if we don't do this
+                    # new .data value size and shape differences will cause
+                    # tracked fakes to produce incorrect guards. This is sound because the TensorVariable
+                    # coming out of set_() below will be a new one, and get
+                    # installed in tracked fakes.
+                    to_remove = []
+                    for tf in tx.output.tracked_fakes:
+                        if tf.source == obj.source:
+                            to_remove.append(tf)
+                    for tf in to_remove:
+                        tx.output.tracked_fakes.remove(tf)
+
+                    # Step 1 - disable grad
+                    version = obj.as_proxy().node.meta["example_value"]._version
+                    with dynamo_disable_grad(tx), torch.no_grad():
+                        # Step 2 - call `set_`
+                        out = wrap_fx_proxy(
+                            tx,
+                            tx.output.create_proxy(
+                                "call_function",
+                                torch.Tensor.set_,
+                                *proxy_args_kwargs([obj, val], {}),
+                            ),
+                        )
+                    # Step 3 - drop the version counter - this is a hack required to get
+                    # .data setting to play correctly with the autograd engine.
+                    if version > 0:
+                        verion = version - 1
+                    tx.output.create_proxy(
+                        "call_function",
+                        torch._C._autograd._unsafe_set_version_counter,
+                        (out.as_proxy(), version),
```

Comment:
I think the generated code here is not quite right. What I will get as output is something like

```
x.set_(...)
unsafe_set_version_counter(x, 10)
```

where 10 is whatever the value of the version counter was *at the time* of tracing. But there's no reason the VC is going to actually be that when you actually run at runtime, because we generally create version counter polymorphic code.

What you really want is to read the VC before the set_ at runtime. This is not so conveniently done when you manually paste the function calls into FX graph. It is much more conveniently done if you introduce a function that does all of this, and solely call that from Dynamo.